File: cpp/tests/test_flow_control.cpp
// 拉取前流控判定的单测 —— 不需要集群。
//
// 为什么必须离线锁死：Java ProcessQueue 的五个阈值里只有**条数**那一条会在"消息小、拉得快"
// 的场景里先命中，其余四条要等真出问题才看得见（队列级字节阈值失效 ⇒ 大消息把堆撑爆；
// 位点跨度失效 ⇒ 队首一条卡住、后面无限堆；topic 级阈值失效 ⇒ 同 topic 多队列各自为政）。
// 判据与 Python `_flow_control_hit` / Rust `flow_control_hit` 逐条同构：
//   1) 条数   >= pullThresholdForQueue（Java 的 max(1, n) 守卫：配 0 也按 1 条算）
//   2) 字节   >= pullThresholdSizeForQueue，单位 **MiB**（<=0 关闭）
//   3) 跨度   **严格大于** consumeConcurrentlyMaxSpan（pending 里 queueOffset 的 max-min；<=0 关闭）
//   4) topic 累计条数 >= pullThresholdForTopic（本实例该 topic **所有**队列合起来；-1 关闭）
//   5) topic 累计字节 >= pullThresholdSizeForTopic，单位 MiB（-1 关闭，且不复用第 2 条的开关）
// 命中一次只记一格 flowControlTriggered()。
#include <cstdio>
#include <limits>
#include <string>
#include <vector>

#include "rocketmq/client/consumer.h"
#include "rocketmq/common/message.h"

using namespace rocketmq;

namespace {

int fails = 0;
int checks = 0;

void expect(bool ok, const std::string& name, const std::string& detail = "") {
    ++checks;
    if (!ok) {
        ++fails;
        std::printf("FAIL %s %s\n", name.c_str(), detail.c_str());
    }
}

constexpr int32_t kMiB = 1024 * 1024;
const char* kTopic = "FlowControlCppUnitTopic";
const char* kOtherTopic = "FlowControlCppUnitOther";
const char* kBroker = "broker-a";

MessageQueue queueOf(const char* topic, int32_t id) {
    return MessageQueue(topic, kBroker, id);
}

MessageExt sized(const char* topic, int32_t storeSize, int64_t queueOffset) {
    MessageExt m;
    m.topic = topic;
    m.brokerName = kBroker;
    m.queueId = 0;
    m.queueOffset = queueOffset;
    m.storeSize = storeSize;
    m.body = "x";
    return m;
}

// 关掉除待测阈值以外的所有闸门：先命中的那条会掩盖被测分支。
void gatesOffExceptCounts(DefaultMQPushConsumer& c) {
    c.setPullThresholdForQueue(std::numeric_limits<int32_t>::max());
    c.setPullThresholdSizeForQueue(0);
    c.setConsumeConcurrentlyMaxSpan(0);
    c.setPullThresholdForTopic(-1);
    c.setPullThresholdSizeForTopic(-1);
}

// 把一批消息预置进某队列的缓冲，并登记该队列已分配（topic 级阈值要靠 mqMap_ 聚合）。
void stage(DefaultMQPushConsumer& c, const MessageQueue& mq, const std::vector<MessageExt>& msgs) {
    const std::string key = DefaultMQPushConsumer::offsetKey(mq);
    c.setAssignedQueue(key, mq);
    c.setPendingMessages(key, msgs);
}

std::vector<MessageExt> nSized(const char* topic, int32_t storeSize, size_t n) {
    std::vector<MessageExt> out;
    for (size_t i = 0; i < n; i++) {
        out.push_back(sized(topic, storeSize, static_cast<int64_t>(i)));
    }
    return out;
}

void testDefaultsOnlyCountGate() {
    DefaultMQPushConsumer c("GID_FlowControlDefaults");
    const MessageQueue mq = queueOf(kTopic, 0);
    // 默认 1000 条 / 100MiB / 跨度 2000，topic 级关闭：小缓冲一律不命中
    stage(c, mq, nSized(kTopic, 100, 1));
    expect(!c.flowControlHit(mq, DefaultMQPushConsumer::offsetKey(mq)), "defaults.noHit");
    expect(c.flowControlTriggered() == 0, "defaults.counterUntouched",
           std::to_string(c.flowControlTriggered()));
    // 缓冲为空（还没拉过任何消息）同样不命中，且跨度不能被算成负数
    stage(c, mq, {});
    expect(!c.flowControlHit(mq, DefaultMQPushConsumer::offsetKey(mq)), "empty.noHit");
}

void testCountGate() {
    DefaultMQPushConsumer c("GID_FlowControlCount");
    const MessageQueue mq = queueOf(kTopic, 0);
    const std::string key = DefaultMQPushConsumer::offsetKey(mq);
    gatesOffExceptCounts(c);
    c.setPullThresholdForQueue(3);
    stage(c, mq, nSized(kTopic, 100, 2));
    expect(!c.flowControlHit(mq, key), "count.below");
    stage(c, mq, nSized(kTopic, 100, 3));
    expect(c.flowControlHit(mq, key), "count.atThreshold");
    expect(c.flowControlTriggered() == 1, "count.counterOnce",
           std::to_string(c.flowControlTriggered()));
    // Java 的 Math.max(1, n) 守卫：配 0 不是"全放行"，而是"1 条就停"
    c.setPullThresholdForQueue(0);
    stage(c, mq, nSized(kTopic, 100, 1));
    expect(c.flowControlHit(mq, key), "count.zeroBehavesAsOne");
}

void testSizeGate() {
    DefaultMQPushConsumer c("GID_FlowControlSize");
    const MessageQueue mq = queueOf(kTopic, 0);
    const std::string key = DefaultMQPushConsumer::offsetKey(mq);
    gatesOffExceptCounts(c);
    c.setPullThresholdSizeForQueue(1);           // 1 MiB
    stage(c, mq, nSized(kTopic, 300, 3));        // 900 B
    expect(!c.flowControlHit(mq, key), "size.below");
    // 单位是 MiB 而不是字节：正好 1 MiB 就算命中（>=）
    stage(c, mq, {sized(kTopic, kMiB, 0)});
    expect(c.flowControlHit(mq, key), "size.exactlyOneMiB");
    stage(c, mq, nSized(kTopic, 2 * kMiB, 4));
    expect(c.flowControlHit(mq, key), "size.over");
    // 0 = 关闭这条闸门，再大的缓冲也不管
    c.setPullThresholdSizeForQueue(0);
    expect(!c.flowControlHit(mq, key), "size.disabled");
}

void testSpanGate() {
    DefaultMQPushConsumer c("GID_FlowControlSpan");
    const MessageQueue mq = queueOf(kTopic, 0);
    const std::string key = DefaultMQPushConsumer::offsetKey(mq);
    gatesOffExceptCounts(c);
    c.setConsumeConcurrentlyMaxSpan(10);
    stage(c, mq, {sized(kTopic, 1, 0), sized(kTopic, 1, 100)});
    expect(c.flowControlHit(mq, key), "span.over");
    // **严格大于**：跨度正好等于阈值不算（Java 同）
    c.setConsumeConcurrentlyMaxSpan(100);
    expect(!c.flowControlHit(mq, key), "span.exactlyAtThreshold");
    // 乱序缓冲也要量出真实跨度（min/max 而不是首尾差）
    c.setConsumeConcurrentlyMaxSpan(10);
    stage(c, mq, {sized(kTopic, 1, 50), sized(kTopic, 1, 5), sized(kTopic, 1, 7)});
    expect(c.flowControlHit(mq, key), "span.unorderedBuffer");
    c.setConsumeConcurrentlyMaxSpan(0);
    expect(!c.flowControlHit(mq, key), "span.disabled");
}

void testTopicCountGate() {
    DefaultMQPushConsumer c("GID_FlowControlTopicCount");
    const MessageQueue q0 = queueOf(kTopic, 0);
    const MessageQueue q1 = queueOf(kTopic, 1);
    const MessageQueue other = queueOf(kOtherTopic, 0);
    gatesOffExceptCounts(c);
    c.setPullThresholdForTopic(2);
    // 单队列 1 条：topic 级也只看到 1 条
    stage(c, q0, nSized(kTopic, 1, 1));
    expect(!c.flowControlHit(q0, DefaultMQPushConsumer::offsetKey(q0)), "topicCount.one");
    // 同 topic 的兄弟队列各 1 条 ⇒ 累计 2 条，两条队列都必须停
    stage(c, q1, nSized(kTopic, 1, 1));
    expect(c.flowControlHit(q0, DefaultMQPushConsumer::offsetKey(q0)), "topicCount.q0");
    expect(c.flowControlHit(q1, DefaultMQPushConsumer::offsetKey(q1)), "topicCount.q1");
    // 别的 topic 不许掺进来：把本 topic 降到 1 条，另一 topic 堆 5 条
    stage(c, q1, {});
    stage(c, other, nSized(kOtherTopic, 1, 5));
    expect(!c.flowControlHit(q0, DefaultMQPushConsumer::offsetKey(q0)),
           "topicCount.otherTopicNotCounted");
}

void testTopicSizeGate() {
    DefaultMQPushConsumer c("GID_FlowControlTopicSize");
    const MessageQueue q0 = queueOf(kTopic, 0);
    const MessageQueue q1 = queueOf(kTopic, 1);
    gatesOffExceptCounts(c);
    // 队列级字节闸门**关掉**（0），只留 topic 级：Rust 曾误用队列级开关当闸门，会让这条静默失效
    c.setPullThresholdSizeForTopic(1);
    stage(c, q0, {sized(kTopic, kMiB / 2, 0)});
    expect(!c.flowControlHit(q0, DefaultMQPushConsumer::offsetKey(q0)), "topicSize.halfMiB");
    stage(c, q1, {sized(kTopic, 3 * kMiB / 2, 0)});
    expect(c.flowControlHit(q0, DefaultMQPushConsumer::offsetKey(q0)), "topicSize.aggregated");
    expect(c.flowControlHit(q1, DefaultMQPushConsumer::offsetKey(q1)), "topicSize.aggregatedQ1");
    // 队列级那道还开着时先命中队列级（判定顺序：条数 → 字节 → 跨度 → topic 条数 → topic 字节）
    c.setPullThresholdSizeForQueue(1);
    stage(c, q0, {sized(kTopic, 2 * kMiB, 0)});
    const int64_t before = c.flowControlTriggered();
    expect(c.flowControlHit(q0, DefaultMQPushConsumer::offsetKey(q0)), "queueSizeWinsOrdering");
    expect(c.flowControlTriggered() == before + 1, "oneHitOneCount",
           std::to_string(c.flowControlTriggered()));
}

}  // namespace

int main() {
    testDefaultsOnlyCountGate();
    testCountGate();
    testSizeGate();
    testSpanGate();
    testTopicCountGate();
    testTopicSizeGate();
    std::printf("flow control: %d checks, %d failures\n", checks, fails);
    return fails == 0 ? 0 : 1;
}

File: python/tests/test_flow_control.py
# 拉取前流控判定（Java ProcessQueue 五个阈值）单测 —— 不需要集群。
#
# 为什么必须离线锁死：五条闸门里只有**条数**那条会在"消息小、拉得快"的场景先命中，
# 其余四条判错方向是**静默**的：
#   - 队列级字节闸门失效 ⇒ 大消息把已拉未消费的堆撑爆（真机表现为 OOM/RT 飙升，无异常）；
#   - 位点跨度闸门失效 ⇒ 队首一条一直消费失败、后面无限堆着，位点跨度失控；
#   - topic 级闸门失效 ⇒ 同 topic 多队列各自为政，单队列都"没超"但实例总量爆掉。
# 单位也只有一个坑：size 阈值是 **MiB** 不是字节；跨度是**严格大于**而条数/字节是 **>=**。
#
# 本文件锁的是**参考实现**本身（Python 是四语言的对齐基准），C++/Rust/.NET 的
# test_flow_control 与这里逐条同构。真机侧"闸门命中之后消息一条都不丢"见
# ../verify_flow_control_live.py。
from rocketmq.client.consumer import DefaultMQPushConsumer
from rocketmq.common.message import MessageExt, MessageQueue

TOPIC = "FlowControlPyUnitTopic"
OTHER_TOPIC = "FlowControlPyUnitOther"
BROKER = "broker-a"
MIB = 1024 * 1024


def _queue(topic=TOPIC, queue_id=0):
    return MessageQueue(topic=topic, broker_name=BROKER, queue_id=queue_id)


def _sized(topic, store_size, queue_offset):
    m = MessageExt(topic=topic, body=b"x")
    m.broker_name = BROKER
    m.store_size = store_size
    m.queue_offset = queue_offset
    return m


def _n_sized(topic, store_size, n):
    return [_sized(topic, store_size, i) for i in range(n)]


def _consumer(group):
    """未 Start 的消费者：_pending/_mq_map 由测试自己预置，不碰网络。"""
    return DefaultMQPushConsumer(group)


def _gates_off(c):
    """关掉除待测阈值以外的所有闸门：先命中的那条会掩盖被测分支。"""
    c.pull_threshold_for_queue = 2 ** 31 - 1
    c.pull_threshold_size_for_queue = 0
    c.consume_concurrently_max_span = 0
    c.pull_threshold_for_topic = -1
    c.pull_threshold_size_for_topic = -1


def _stage(c, mq, msgs):
    """预置某队列的缓冲 + 登记为已分配（topic 级阈值靠 _mq_map 聚合兄弟队列）。"""
    key = c._mq_key(mq)
    c._mq_map[key] = mq
    c._pending[key] = list(msgs)
    return key


def _hit(c, mq, key=None):
    return c._flow_control_hit(mq, key or c._mq_key(mq))


def test_defaults_only_the_count_gate_is_live():
    c = _consumer("G_fc_defaults")
    mq = _queue()
    # 默认 1000 条 / 100MiB / 跨度 2000，topic 级关闭：小缓冲一律不命中
    _stage(c, mq, _n_sized(TOPIC, 100, 1))
    assert not _hit(c, mq)
    assert c._flow_control_triggered == 0
    # 缓冲为空（还没拉过任何消息）同样不命中，且跨度不能被算成负数
    _stage(c, mq, [])
    assert not _hit(c, mq)


def test_count_gate_fires_at_threshold_and_treats_zero_as_one():
    c = _consumer("G_fc_count")
    mq = _queue()
    _gates_off(c)
    c.pull_threshold_for_queue = 3
    _stage(c, mq, _n_sized(TOPIC, 100, 2))
    assert not _hit(c, mq)
    _stage(c, mq, _n_sized(TOPIC, 100, 3))
    assert _hit(c, mq)
    assert c._flow_control_triggered == 1
    # Java 的 Math.max(1, n) 守卫：配 0 不是"全放行"，而是"1 条就停"
    c.pull_threshold_for_queue = 0
    _stage(c, mq, _n_sized(TOPIC, 100, 1))
    assert _hit(c, mq), "pull_threshold_for_queue=0 要按 1 条算（Java 的 max(1,n) 守卫）"


def test_size_gate_is_in_mib_and_disabled_by_zero():
    c = _consumer("G_fc_size")
    mq = _queue()
    _gates_off(c)
    c.pull_threshold_size_for_queue = 1          # 1 MiB
    _stage(c, mq, _n_sized(TOPIC, 300, 3))       # 900 B
    assert not _hit(c, mq)
    # 单位是 MiB 而不是字节：正好 1 MiB 就算命中（>=）
    _stage(c, mq, [_sized(TOPIC, MIB, 0)])
    assert _hit(c, mq), "正好 1MiB 要命中（>=，不是 >）"
    _stage(c, mq, _n_sized(TOPIC, 2 * MIB, 4))
    assert _hit(c, mq)
    # 0 = 关闭这条闸门，再大的缓冲也不管
    c.pull_threshold_size_for_queue = 0
    assert not _hit(c, mq)


def test_span_gate_is_strictly_greater_and_measures_real_span():
    c = _consumer("G_fc_span")
    mq = _queue()
    _gates_off(c)
    c.consume_concurrently_max_span = 10
    _stage(c, mq, [_sized(TOPIC, 1, 0), _sized(TOPIC, 1, 100)])
    assert _hit(c, mq)
    # **严格大于**：跨度正好等于阈值不算（Java 同）
    c.consume_concurrently_max_span = 100
    assert not _hit(c, mq), "跨度 100 不严格大于 100，不该命中"
    # 乱序缓冲也要量出真实跨度（min/max 而不是首尾差）
    c.consume_concurrently_max_span = 10
    _stage(c, mq, [_sized(TOPIC, 1, 50), _sized(TOPIC, 1, 5), _sized(TOPIC, 1, 7)])
    assert _hit(c, mq), "乱序缓冲要按 max-min 量跨度，不是首尾差"
    c.consume_concurrently_max_span = 0
    assert not _hit(c, mq)


def test_topic_count_gate_aggregates_siblings_but_not_other_topics():
    c = _consumer("G_fc_topic_count")
    q0, q1 = _queue(queue_id=0), _queue(queue_id=1)
    other = _queue(topic=OTHER_TOPIC)
    _gates_off(c)
    c.pull_threshold_for_topic = 2
    # 单队列 1 条：topic 级也只看到 1 条
    _stage(c, q0, _n_sized(TOPIC, 1, 1))
    assert not _hit(c, q0)
    # 同 topic 的兄弟队列各 1 条 ⇒ 累计 2 条，两条队列都必须停
    k1 = _stage(c, q1, _n_sized(TOPIC, 1, 1))
    assert _hit(c, q0)
    assert _hit(c, q1, k1)
    # 别的 topic 不许掺进来：把本 topic 降到 1 条，另一 topic 堆 5 条
    _stage(c, q1, [])
    _stage(c, other, _n_sized(OTHER_TOPIC, 1, 5))
    assert not _hit(c, q0), "别的 topic 的缓冲不能算进本 topic 的累计"


def test_topic_size_gate_has_its_own_switch_and_runs_last():
    c = _consumer("G_fc_topic_size")
    q0, q1 = _queue(queue_id=0), _queue(queue_id=1)
    _gates_off(c)
    # 队列级字节闸门**关掉**（0），只留 topic 级：误用队列级开关当闸门会让这条静默失效
    c.pull_threshold_size_for_topic = 1
    _stage(c, q0, [_sized(TOPIC, MIB // 2, 0)])
    assert not _hit(c, q0)
    _stage(c, q1, [_sized(TOPIC, 3 * MIB // 2, 0)])
    assert _hit(c, q0), "队列级 size 闸门关掉时主题级 size 仍要生效"
    assert _hit(c, q1)
    # 队列级那道还开着时先命中队列级（判定顺序：条数 → 字节 → 跨度 → topic 条数 → topic 字节）
    c.pull_threshold_size_for_queue = 1
    _stage(c, q0, [_sized(TOPIC, 2 * MIB, 0)])
    before = c._flow_control_triggered
    assert _hit(c, q0)
    # 一次判定只记一格，即使多条闸门同时命中
    assert c._flow_control_triggered == before + 1


def test_one_hit_counts_one_tick_even_when_every_gate_matches():
    """计数器是"因流控停了几次"的观测面：五条闸门同时命中也只能加一格。"""
    c = _consumer("G_fc_counter")
    q0, q1 = _queue(queue_id=0), _queue(queue_id=1)
    _gates_off(c)
    c.pull_threshold_for_queue = 1
    c.pull_threshold_size_for_queue = 1
    c.consume_concurrently_max_span = 1
    c.pull_threshold_for_topic = 1
    c.pull_threshold_size_for_topic = 1
    # 2 条 2MiB、位点跨到 100 的缓冲：五条闸门同时命中
    _stage(c, q0, [_sized(TOPIC, 2 * MIB, 0), _sized(TOPIC, 2 * MIB, 100)])
    _stage(c, q1, [_sized(TOPIC, 2 * MIB, 0)])
    assert _hit(c, q0)
    assert c._flow_control_triggered == 1
